packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts

```diff
@@ -0,0 +1,178 @@
+          const scriptsToOptimize = [];
+
+          // Analyze the compilation assets for scripts that require optimization
+          for (const assetName of Object.keys(compilationAssets)) {
+            if (assetName.endsWith('.js')) {
+              const scriptAsset = compilation.getAsset(assetName);
+              if (scriptAsset && !scriptAsset.info.minimized) {
+                const { source, map } = scriptAsset.source.sourceAndMap();
+                scriptsToOptimize.push({
+                  name: scriptAsset.name,
+                  code: typeof source === 'string' ? source : source.toString(),
+                  map,
+                });
+              }
+            }
+          }
+
+          if (scriptsToOptimize.length === 0) {
+            return;
+          }
+
+          // Ensure all replacement values are strings which is the expected type for esbuild
+          let define: Record<string, string> | undefined;
+          if (this.options.define) {
+            define = {};
+            for (const [key, value] of Object.entries(this.options.define)) {
+              define[key] = String(value);
+            }
+          }
+
+          // Setup the options used by all worker tasks
+          const optimizeOptions = {
+            sourcemap: this.options.sourcemap,
+            define,
+            keepNames: this.options.keepNames,
+            target: this.options.target ? Number(this.options.target.slice(2)) : 2017,
+            removeLicenses: this.options.removeLicenses,
+            advanced: this.options.advanced,
+          };
+
+          // Sort scripts so larger scripts start first - worker pool uses a FIFO queue
+          scriptsToOptimize.sort((a, b) => a.code.length - b.code.length);
+
+          // Initialize the task worker pool
+          const workerPath = require.resolve('./javascript-optimizer-worker');
+          const workerPool = new Piscina({
+            filename: workerPath,
+            maxThreads: MAX_OPTIMIZE_WORKERS,
+          });
+
+          // Enqueue script optimization tasks and update compilation assets as the tasks complete
+          try {
+            const tasks = [];
+            for (const { name, code, map } of scriptsToOptimize) {
+              tasks.push(
+                workerPool
+                  .run({
+                    asset: {
+                      name,
+                      code,
+                      map,
+                    },
+                    options: optimizeOptions,
+                  })
+                  .then(
+                    ({ code, name, map }) => {
+                      let optimizedAsset;
+                      if (map) {
+                        optimizedAsset = new SourceMapSource(code, name, map);
+                      } else {
+                        optimizedAsset = new OriginalSource(code, name);
+                      }
+                      compilation.updateAsset(name, optimizedAsset, { minimized: true });
```

Comment:
It would be great if in a followup we cache these results using Webpack's caching API.

Reply:
Yes. There is already a decent amount of new code in this PR so I didn't want to add the complexity of the Webpack 5 caching initially. Adding support for caching would be a fairly straightforward incremental improvement once this PR lands.